src/modules/70-pipeline/components/PipelineStudio/PipelineVariables/PipelineVariables.tsx

```diff
@@ -146,23 +127,117 @@ export function PipelineVariablesWithRef(
+      ) : (
+        <div className={css.content}>
+          <VariablesHeader applyChanges={applyChanges} discardChanges={discardChanges} />
+          <div className={css.variableList} ref={pipelineVariablesRef as any}>
+            <GitSyncStoreProvider>
+              <NestedAccordionPanel
+                noAutoScroll
+                isDefaultOpen
+                key="pipeline"
+                id="pipeline"
+                addDomId
+                collapseProps={{
+                  keepChildrenMounted: true
+                }}
+                summary={<VariableAccordionSummary>{getString('common.pipeline')}</VariableAccordionSummary>}
+                summaryClassName={css.stageSummary}
+                detailsClassName={css.pipelineDetails}
+                panelClassName={css.pipelineMarginBottom}
+                details={
+                  <PipelineCardPanel
+                    variablePipeline={variablesPipeline}
+                    originalPipeline={originalPipeline}
+                    pipeline={pipeline}
+                    stepsFactory={stepsFactory}
+                    updatePipeline={updatePipeline}
+                    metadataMap={metadataMap}
+                    readonly={isReadonly || !!pipeline.template}
+                    allowableTypes={allowableTypes}
+                  />
+                }
+              />
+            </GitSyncStoreProvider>
+          </div>
+        </div>
+      )}
+      <ConfirmationDialog
+        titleText={getString('variablesText')}
+        contentText={getString('pipeline.stepConfigHasChanges')}
+        isOpen={isConfirmationDialogOpen}
+        confirmButtonText={getString('applyChanges')}
+        cancelButtonText={getString('cancel')}
+        onClose={handleConfirmation}
+        intent={Intent.WARNING}
+      />
+    </div>
+  )
+}
+
+export interface PipelineCardPanelProps extends PipelineCardProps {
+  originalPipeline: PipelineInfoConfig
+}
+
+export function PipelineCardPanel(props: PipelineCardPanelProps): React.ReactElement {
+  const {
+    variablePipeline,
+    pipeline,
+    originalPipeline,
+    stepsFactory,
+    metadataMap,
+    allowableTypes,
+    updatePipeline,
+    readonly
+  } = props
+
+  const updateStages = React.useCallback(
+    (newStage: StageElementConfig, stages: StageElementWrapperConfig[]): StageElementWrapperConfig[] => {
+      return stages.map(node => {
+        if (node.stage?.identifier === newStage.identifier) {
+          return { stage: newStage }
+        } else if (node.parallel) {
+          return {
+            parallel: updateStages(newStage, node.parallel)
+          }
+        }
+        return node
+      })
+    },
+    []
+  )
+
+  const updateStage = React.useCallback(
+    async (values: StageElementConfig) => {
+      if (pipeline.stages) {
+        set(pipeline, 'stages', updateStages(values, pipeline.stages))
+        updatePipeline(pipeline)
+      }
+    },
+    [pipeline, updatePipeline]
```

Comment:
Codacy has a fix for the issue: [React Hook React.useCallback has a missing dependency: 'updateStages'. Either include it or remove the dependency array.](https://app.codacy.com/gh/harness/harness-core-ui/pullRequest?prid=9232314)
```suggestion
    [pipeline, updatePipeline, updateStages]
```